File: src/main/java/com/tososomaru/language/scl/SCLParserDefinition.java
package com.tososomaru.language.scl;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import com.tososomaru.language.scl.parser.SCLParser;
import com.tososomaru.language.scl.psi.SCLFile;
import com.tososomaru.language.scl.psi.SCLTypes;
import org.jetbrains.annotations.NotNull;

public class SCLParserDefinition implements ParserDefinition {

    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(SCLTypes.COMMENT);

    public static final IFileElementType FILE = new IFileElementType(SCLLanguage.INSTANCE);

    @Override
    public @NotNull Lexer createLexer(Project project) {
        return new SCLFlexAdapter();
    }

    @Override
    public @NotNull PsiParser createParser(Project project) {
        return new SCLParser();
    }

    @Override
    public @NotNull IFileElementType getFileNodeType() {
        return FILE;
    }

    @Override
    public @NotNull TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @Override
    public @NotNull TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @Override
    public @NotNull PsiElement createElement(ASTNode node) {
        return SCLTypes.Factory.createElement(node);
    }

    @Override
    public @NotNull PsiFile createFile(@NotNull FileViewProvider viewProvider) {
        return new SCLFile(viewProvider);
    }

    @Override
    public @NotNull TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @Override
    public @NotNull SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
        return SpaceRequirements.MAY;
    }

}

File: src/main/kotlin/com/tososomaru/language/scl/SCLParserDefinition.kt
package com.tososomaru.language.scl

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.ParserDefinition.SpaceRequirements
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import com.tososomaru.language.scl.parser.SCLParser
import com.tososomaru.language.scl.psi.SCLFile
import com.tososomaru.language.scl.psi.SCLTypes

class SCLParserDefinition : ParserDefinition {
    override fun createLexer(project: Project): Lexer = SCLLexerAdapter()

    override fun createParser(project: Project): PsiParser = SCLParser()

    override fun getFileNodeType(): IFileElementType = FILE

    override fun getCommentTokens(): TokenSet = COMMENTS

    override fun getStringLiteralElements(): TokenSet = TokenSet.EMPTY

    override fun createElement(node: ASTNode): PsiElement = SCLTypes.Factory.createElement(node)

    override fun createFile(viewProvider: FileViewProvider): PsiFile = SCLFile(viewProvider)

    override fun getWhitespaceTokens(): TokenSet = WHITE_SPACES

    override fun spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode): SpaceRequirements {
        return SpaceRequirements.MAY
    }

    companion object {
        val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
        val COMMENTS = TokenSet.create(SCLTypes.COMMENT)
        val FILE = IFileElementType(SCLLanguage)
    }
}